proto/src/lib.rs

```diff
@@ -344,14 +344,79 @@ pub mod logger {
         type Item = LogItem;
 
         async fn receive(&mut self, items: Vec<Self::Item>) {
+            // A log vector should never be received without any items. We clone the first item
+            // here so we can use IDs to generate a rate limiting logline, which we will send to
+            // the logger as a warning to the user that they are being rate limited.
+            let Some(item) = items.first().cloned() else {
+                error!("received log vector without any items");
+
+                return;
+            };
```

Comment:
`items::is_empty()`?

Reply:
We need the item here (`let Some(item)`) in case we want to send an item to the logger about the caller being rate limited.